### PR TITLE
Use RwLock for active collectors collection 

### PR DESCRIPTION
**Pull request by TheCataliasTNT2k**
One of the main functions used a Mutex to guard a shared variable.
This may not impact small bots, but can significantly slow down bots, which have a huge volume of GatewayEvents.
This Pullrequest replaces these Mutexes with a RwLock, which allows multiple simultaneous readers.
This should not break any existing code.

### CODE REVIEW

**Comment by GnomedDev** [comment]
This is a breaking change, please re-target to the next branch.

**Comment by TheCataliasTNT2k** [comment]
I do not know, what this should break, I tested it with current and it worked without any other additions, but I changed it.

**Comment by GnomedDev** [comment]
```rs
let mutex = Arc::new(tokio::sync::Mutex::new("".to_string()));
Shard::new(mutex, "", shard_info, GatewayIntents::all(), None).await?;
```

This is an example of user code that would break if this PR was merged. Breaking changes means *any* public API change that stops compilation, with few exceptions. This PR now needs rebasing to drop the extra commits from current and to solve the conflicts.

**Comment by TheCataliasTNT2k** [comment]
Ah, I forgot about the ws_url, I see. I will hust rebase it onto next, this should be enough.

**Comment by TheCataliasTNT2k** [comment]
I now got a lot of "Unresolved reference: `instrument`", but I ignored them, because they already existed on the branch.

**Comment by GnomedDev** [comment]
Disable the tracing_instrument feature in your builds, it is currently broken, that will get those errors to go away.

**Comment by TheCataliasTNT2k** [comment]
I did not get build errors, `cargo test --features full` runs successfully, jut my IDE is very unhappy.

**Comment by GnomedDev** [comment]
Yes, but you have probably enabled `--all-features` in your IDE, so it enables the broken feature.

File: src/gateway/bridge/shard_runner.rs [hunk 1]
```diff
@@ -42,8 +42,9 @@ pub struct ShardRunner {
     #[cfg(feature = "cache")]
     pub cache: Arc<Cache>,
     pub http: Arc<Http>,
+    // do not use mutex! this is used for almost every gateway event!
```
Comment (GnomedDev) [inline src/gateway/bridge/shard_runner.rs]: Remove this comment, if there is a reason someone PRs this back to Mutex it will have to justify the performance hit anyway.

**Comment by TheCataliasTNT2k** [comment]
I will let rustfmt fix the linter issue.
What to do about the other linter issue? Should I really create a new type alias for that?

File: src/gateway/bridge/shard_runner.rs [hunk 2]
```diff
@@ -171,7 +172,14 @@ impl ShardRunner {
 
             if let Some(event) = event {
                 #[cfg(feature = "collector")]
-                self.collectors.lock().expect("poison").retain_mut(|callback| (callback.0)(&event));
+                {
+                    let read_lock = self.collectors.read();
+                    let to_remove: Vec<_> = read_lock.iter().filter(|callback| !callback.0(&event)).cloned().collect();
+                    drop(read_lock);
+                    if !to_remove.is_empty() {
+                        self.collectors.write().retain(|f| !to_remove.contains(f));
```
Comment (GnomedDev) [inline src/gateway/bridge/shard_runner.rs:183]: What happens if a collector gets pushed into this Vec between the read lock being dropped and the write lock being acquired?

File: src/gateway/bridge/mod.rs [hunk 3]
```diff
@@ -97,9 +98,16 @@ pub struct ShardRunnerInfo {
...
+pub struct CollectorCallback(pub Arc<Box<dyn Fn(&Event) -> bool + Send + Sync>>);
 impl std::fmt::Debug for CollectorCallback {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_tuple("CollectorCallback").finish()
     }
 }
+
+impl PartialEq for CollectorCallback {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::as_ptr(&self.0) == Arc::as_ptr(&other.0)
```
Comment (GnomedDev) [inline src/gateway/bridge/mod.rs]: Arc has a method for this, `Arc::ptr_eq`.

File: src/gateway/bridge/mod.rs [hunk 4]
```diff
@@ -97,9 +98,16 @@ pub struct ShardRunnerInfo {
 /// Newtype around a callback that will be called on every incoming request. As long as this
 /// collector should still receive events, it should return `true`. Once it returns `false`, it is
 /// removed.
-pub struct CollectorCallback(pub Box<dyn Fn(&Event) -> bool + Send + Sync>);
+#[derive(Clone)]
+pub struct CollectorCallback(pub Arc<Box<dyn Fn(&Event) -> bool + Send + Sync>>);
```
Comment (GnomedDev) [inline src/gateway/bridge/mod.rs]: Arc does not require T to be sized, so you can store `Arc<dyn Fn(&Event) + ...>`

**Review by GnomedDev — CHANGES_REQUESTED** [review]

File: src/gateway/bridge/mod.rs (hunk 4 above)
Comment (GnomedDev) [reply src/gateway/bridge/mod.rs]: This should incidentally solve your type complexity lint issue.

File: src/gateway/bridge/shard_runner.rs (hunk 2 above)
Comment (TheCataliasTNT2k) [reply src/gateway/bridge/shard_runner.rs:183]: Nothing, the collector will not be run. But the code should not delete "wrong" collectors.

File: src/gateway/bridge/mod.rs (hunk 3 above)
Comment (TheCataliasTNT2k) [reply src/gateway/bridge/mod.rs]: It has, but the conversion does not work as expected. This method failed my tests.

File: src/gateway/bridge/shard_runner.rs (hunk 2 above)
Comment (GnomedDev) [reply src/gateway/bridge/shard_runner.rs:183]: Okay, might be worth adding a comment here as it looks mighty suspicious without looking into it.

File: src/gateway/bridge/mod.rs (hunk 4 above)
Comment (TheCataliasTNT2k) [reply src/gateway/bridge/mod.rs]: Will update that.

File: src/gateway/bridge/shard_runner.rs (hunk 1 above)
Comment (TheCataliasTNT2k) [reply src/gateway/bridge/shard_runner.rs]: Ok, I will remove it.

File: src/collector.rs [hunk 5]
```diff
@@ -1,3 +1,8 @@
+// Or we'll get deprecation warnings from our own deprecated type (seriously Rust?)
+#![allow(deprecated)]
```
Comment (GnomedDev) [inline src/collector.rs]: Why is this here?

File: src/gateway/bridge/mod.rs (hunk 3 above)
Comment (GnomedDev) [reply src/gateway/bridge/mod.rs]: I very much doubt that... https://doc.rust-lang.org/stable/src/alloc/sync.rs.html#1786

File: src/gateway/bridge/shard_runner.rs (hunk 2 above)
Comment (TheCataliasTNT2k) [reply src/gateway/bridge/shard_runner.rs:183]: I did that

File: src/collector.rs (hunk 5 above)
Comment (TheCataliasTNT2k) [reply src/collector.rs]: This was not added by me, I thought, it was in the Git history somewhere, but I can not find the commit. I will remove it then...

File: src/gateway/bridge/mod.rs (hunk 3 above)
Comment (TheCataliasTNT2k) [reply src/gateway/bridge/mod.rs]: I did not realise, this fas fixed, because it was a problem some time ago: https://github.com/rust-lang/rust/issues/103763
I will change it.

File: src/collector.rs (hunk 5 above)
Comment (TheCataliasTNT2k) [reply src/collector.rs]: I guess, it was added during rebase onto next yesterday by some of the commits.

**Comment by TheCataliasTNT2k** [comment]
Too stupid to see errors sometimes...

**Review by GnomedDev — APPROVED** [review]
Looks good to me, thanks for the PR!

**Comment by TheCataliasTNT2k** [comment]
Thanks for the help!
